bundles/org.openhab.core.model.yaml/src/main/java/org/openhab/core/model/yaml/internal/semantics/YamlSemanticTagProvider.java

```diff
@@ -0,0 +1,125 @@
+/**
+ * Copyright (c) 2010-2023 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.core.model.yaml.internal.semantics;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.core.common.registry.AbstractProvider;
+import org.openhab.core.model.yaml.YamlElement;
+import org.openhab.core.model.yaml.YamlFile;
+import org.openhab.core.model.yaml.YamlModelListener;
+import org.openhab.core.semantics.SemanticTag;
+import org.openhab.core.semantics.SemanticTagImpl;
+import org.openhab.core.semantics.SemanticTagProvider;
+import org.openhab.core.semantics.SemanticTagRegistry;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link YamlSemanticTagProvider} is an OSGi service, that allows to define semantic tags
+ * in YAML configuration files in folder conf/tags.
+ * Files can be added, updated or removed at runtime.
+ * These semantic tags are automatically exposed to the {@link SemanticTagRegistry}.
+ *
+ * @author Laurent Garnier - Initial contribution
+ */
+@NonNullByDefault
+@Component(immediate = true, service = { SemanticTagProvider.class, YamlSemanticTagProvider.class,
+        YamlModelListener.class })
+public class YamlSemanticTagProvider extends AbstractProvider<SemanticTag>
+        implements SemanticTagProvider, YamlModelListener<YamlSemanticTag> {
+
+    private final Logger logger = LoggerFactory.getLogger(YamlSemanticTagProvider.class);
+
+    private final List<SemanticTag> tags = new ArrayList<>();
```

Comment:
It is my understanding that `SemanticTag`s are unique and must not have duplicates. Correct? In that case I would suggest to use a `TreeSet` instead. That has two advantages: it is automatically sorted (because elements are put in the correct place when they are added) and it also prevents duplicates.

Reply:
My provider must return the elements (method getAll) in a particular order, sorted by semantic tag UID.
If a TreeSet allows that, why not.
I should investigate and test.

Reply:
Should work. Done.